plugins/out_forward/forward.c

```diff
@@ -943,25 +979,53 @@ static int flush_forward_mode(struct flb_forward *ctx,
     /* Tag */
     flb_forward_format_append_tag(ctx, fc, &mp_pck, NULL, tag, tag_len);
 
-    entries = flb_mp_count(data, bytes);
-    msgpack_pack_array(&mp_pck, entries);
+    if (fc->compress == COMPRESS_GZIP) {
+        /* When compress is set, we switch from using Forward mode to using
+         * CompressedPackedForward mode.
+         */
+        ret = flb_gzip_compress((void *) data, bytes, &final_data, &final_bytes);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "could not compress entries");
+            msgpack_sbuffer_destroy(&mp_sbuf);
+            return FLB_RETRY;
+        }
+
+        msgpack_pack_bin(&mp_pck, final_bytes);
+
+    } else {
+        final_data = (void *) data;
+        final_bytes = bytes;
+
+        entries = flb_mp_count(data, bytes);
+        msgpack_pack_array(&mp_pck, entries);
+    }
 
     /* Write message header */
     ret = flb_io_net_write(u_conn, mp_sbuf.data, mp_sbuf.size, &bytes_sent);
     if (ret == -1) {
         flb_plg_error(ctx->ins, "could not write forward header");
         msgpack_sbuffer_destroy(&mp_sbuf);
+        if (final_data != data) {
```

Comment:
Probably `if (fc->compress == COMPRESS_GZIP) {` is better as
a conditional expression here.